# Global Wheat Detection

This is the training notebook to my first solution for taking part in the ["Global Wheat Detection"](https://www.kaggle.com/c/global-wheat-detection) competition on Kaggle. 

The goal of this competition was using object detection to detect wheat heads in pictures to simplify and automate some parts of agricultural research. The competition was hosted by the [Global Institute for Food Security](https://www.gifs.ca/) of the [University of Saskatchewan](http://www.usask.ca/) among others. Previous models had always worked very well on wheat populations in certain parts of the world, while performing poorly on others. The ultimate aim of the competition was to create models without this geographical bias.

This solution is based on the newly released YOLOv5 algorithm.

## Setup
The general goal of this notebook will be that it can be run both on my machine, as well as on kaggle. 

### Making the filesystem

Yolo needs a fairly specific file system to work, so I will now make the directories necessary

#### Detecting the operating system

In [1]:
import os
system = os.name
if system == 'posix':
    kaggle = True
    windows = False
    print('running on kaggle')
elif system == 'nt':
    kaggle = False
    windows = True
    print('running on windows')
else:
    print('unknown system')

running on windows


Making the folder for the training pictures and copying them from the ../input directory

In [3]:
if kaggle:
    !mkdir data
    !mkdir data/images
    !cp -r ../input/global-wheat-detection/train ./data/images/
    !cp -r ../input/global-wheat-detection/test ./data/images/
elif windows:
    !mkdir data
    !mkdir data\images
    !mkdir data\images\train
    !copy ..\input\global-wheat-detection\train\*.jpg .\data\images\train
    !mkdir data\images\test
    !copy ..\input\global-wheat-detection\test\*.jpg .\data\images\test
    

..\input\global-wheat-detection\train\00333207f.jpg
..\input\global-wheat-detection\train\005b0d8bb.jpg
..\input\global-wheat-detection\train\006a994f7.jpg
..\input\global-wheat-detection\train\00764ad5d.jpg
..\input\global-wheat-detection\train\00b5c6764.jpg
..\input\global-wheat-detection\train\00b5fefed.jpg
..\input\global-wheat-detection\train\00b70a919.jpg
..\input\global-wheat-detection\train\00e903abe.jpg
..\input\global-wheat-detection\train\00ea5e5ee.jpg
..\input\global-wheat-detection\train\010b216d4.jpg
..\input\global-wheat-detection\train\010c93b99.jpg
..\input\global-wheat-detection\train\010dbcc8f.jpg
..\input\global-wheat-detection\train\0114c88aa.jpg
..\input\global-wheat-detection\train\01189a3c3.jpg
..\input\global-wheat-detection\train\0126b7d11.jpg
..\input\global-wheat-detection\train\013669953.jpg
..\input\global-wheat-detection\train\01397a84c.jpg
..\input\global-wheat-detection\train\013fd7d80.jpg
..\input\global-wheat-detection\train\015939012.jpg
..\input\glo

..\input\global-wheat-detection\train\0e86615d0.jpg
..\input\global-wheat-detection\train\0e90d45c6.jpg
..\input\global-wheat-detection\train\0eab04c31.jpg
..\input\global-wheat-detection\train\0eab1baab.jpg
..\input\global-wheat-detection\train\0ebac9a23.jpg
..\input\global-wheat-detection\train\0ee812df9.jpg
..\input\global-wheat-detection\train\0eed0a689.jpg
..\input\global-wheat-detection\train\0f0bb7843.jpg
..\input\global-wheat-detection\train\0f43bb6d7.jpg
..\input\global-wheat-detection\train\0f47f54f1.jpg
..\input\global-wheat-detection\train\0f4911203.jpg
..\input\global-wheat-detection\train\0f5267442.jpg
..\input\global-wheat-detection\train\0f9100915.jpg
..\input\global-wheat-detection\train\0f9717e87.jpg
..\input\global-wheat-detection\train\0fb8ec104.jpg
..\input\global-wheat-detection\train\0fcdc222b.jpg
..\input\global-wheat-detection\train\0fd4ca0d4.jpg
..\input\global-wheat-detection\train\0fd63ed98.jpg
..\input\global-wheat-detection\train\0ff60b2bc.jpg
..\input\glo

..\input\global-wheat-detection\train\2aa35dade.jpg
..\input\global-wheat-detection\train\2ac5f9e85.jpg
..\input\global-wheat-detection\train\2ad7fa68e.jpg
..\input\global-wheat-detection\train\2adceacda.jpg
..\input\global-wheat-detection\train\2ae1f9d88.jpg
..\input\global-wheat-detection\train\2ae9c276f.jpg
..\input\global-wheat-detection\train\2b1986944.jpg
..\input\global-wheat-detection\train\2b2612efa.jpg
..\input\global-wheat-detection\train\2b32a4cc4.jpg
..\input\global-wheat-detection\train\2b3c77755.jpg
..\input\global-wheat-detection\train\2b45c0165.jpg
..\input\global-wheat-detection\train\2b5e0f1a2.jpg
..\input\global-wheat-detection\train\2b7b5621c.jpg
..\input\global-wheat-detection\train\2b99d68df.jpg
..\input\global-wheat-detection\train\2bc10136f.jpg
..\input\global-wheat-detection\train\2bd2f0714.jpg
..\input\global-wheat-detection\train\2bfadaa91.jpg
..\input\global-wheat-detection\train\2c28cb6dd.jpg
..\input\global-wheat-detection\train\2c2c2a3d6.jpg
..\input\glo

..\input\global-wheat-detection\train\5f8ec0456.jpg
..\input\global-wheat-detection\train\5f9151e33.jpg
..\input\global-wheat-detection\train\5f9cfe156.jpg..\input\global-wheat-detection\test\aac893a91.jpg
..\input\global-wheat-detection\test\cb8d261a3.jpg

..\input\global-wheat-detection\test\cc3532ff6.jpg
..\input\global-wheat-detection\train\5fab819b3.jpg
..\input\global-wheat-detection\test\f5a1f0358.jpg
..\input\global-wheat-detection\train\5fabe93b1.jpg
       10 file(s) copied.
..\input\global-wheat-detection\train\5fb074443.jpg
..\input\global-wheat-detection\train\5fdc74081.jpg
..\input\global-wheat-detection\train\60044b7cb.jpg
..\input\global-wheat-detection\train\602e4b162.jpg
..\input\global-wheat-detection\train\60870e668.jpg
..\input\global-wheat-detection\train\60913f4db.jpg
..\input\global-wheat-detection\train\609c993a9.jpg
..\input\global-wheat-detection\train\60b440046.jpg
..\input\global-wheat-detection\train\60cb22568.jpg
..\input\global-wheat-detection\train\60fc

..\input\global-wheat-detection\train\7939edac8.jpg
..\input\global-wheat-detection\train\793ef2974.jpg
..\input\global-wheat-detection\train\7969968d8.jpg
..\input\global-wheat-detection\train\797dd3e01.jpg
..\input\global-wheat-detection\train\798a261f0.jpg
..\input\global-wheat-detection\train\79964dffd.jpg
..\input\global-wheat-detection\train\79a1d0dad.jpg
..\input\global-wheat-detection\train\79adcb57c.jpg
..\input\global-wheat-detection\train\79b825776.jpg
..\input\global-wheat-detection\train\79c8a1627.jpg
..\input\global-wheat-detection\train\79ce67248.jpg
..\input\global-wheat-detection\train\79d5a9b6b.jpg
..\input\global-wheat-detection\train\79eca0678.jpg
..\input\global-wheat-detection\train\7a0fe3392.jpg
..\input\global-wheat-detection\train\7a77062fb.jpg
..\input\global-wheat-detection\train\7a9d1d1f2.jpg
..\input\global-wheat-detection\train\7a9d28a53.jpg
..\input\global-wheat-detection\train\7aa682e0f.jpg
..\input\global-wheat-detection\train\7ab4f3778.jpg
..\input\glo

Making the folders to house the labels

In [4]:
if kaggle:
    !mkdir data/labels
    !mkdir data/labels/train
elif windows:
    !mkdir data\labels
    !mkdir data\labels\train

..\input\global-wheat-detection\train\96fa7bbcb.jpg
..\input\global-wheat-detection\train\97316e61d.jpg
..\input\global-wheat-detection\train\97372d371.jpg
..\input\global-wheat-detection\train\974b78929.jpg
..\input\global-wheat-detection\train\975c97e30.jpg
..\input\global-wheat-detection\train\9764d9f54.jpg
..\input\global-wheat-detection\train\976746c7d.jpg
..\input\global-wheat-detection\train\976d1eb84.jpg
..\input\global-wheat-detection\train\97788b3ae.jpg
..\input\global-wheat-detection\train\9780d64f5.jpg
..\input\global-wheat-detection\train\97851b9c6.jpg
..\input\global-wheat-detection\train\9785e4e53.jpg
..\input\global-wheat-detection\train\9790095bb.jpg
..\input\global-wheat-detection\train\97b19f207.jpg
..\input\global-wheat-detection\train\97ba6330a.jpg
..\input\global-wheat-detection\train\97f0e1800.jpg
..\input\global-wheat-detection\train\980c88da5.jpg
..\input\global-wheat-detection\train\9832f05e1.jpg
..\input\global-wheat-detection\train\9858d67dc.jpg
..\input\glo

..\input\global-wheat-detection\train\b13f38e6f.jpg
..\input\global-wheat-detection\train\b1642dc6f.jpg
..\input\global-wheat-detection\train\b1643e118.jpg
..\input\global-wheat-detection\train\b18b34ed2.jpg
..\input\global-wheat-detection\train\b1990fa45.jpg
..\input\global-wheat-detection\train\b1b1b8fba.jpg
..\input\global-wheat-detection\train\b1c1aade8.jpg
..\input\global-wheat-detection\train\b1cf5e43c.jpg
..\input\global-wheat-detection\train\b1e31c10b.jpg
..\input\global-wheat-detection\train\b1e4a723e.jpg
..\input\global-wheat-detection\train\b20d35e4d.jpg
..\input\global-wheat-detection\train\b21ca086c.jpg
..\input\global-wheat-detection\train\b22f456aa.jpg
..\input\global-wheat-detection\train\b24a3ecee.jpg
..\input\global-wheat-detection\train\b2672900e.jpg
..\input\global-wheat-detection\train\b26aeda3c.jpg
..\input\global-wheat-detection\train\b27a68a05.jpg
..\input\global-wheat-detection\train\b28548e32.jpg
..\input\global-wheat-detection\train\b28fb55cc.jpg
..\input\glo

..\input\global-wheat-detection\train\c64a8e237.jpg
..\input\global-wheat-detection\train\c64da3b0d.jpg
..\input\global-wheat-detection\train\c6654ff53.jpg
..\input\global-wheat-detection\train\c66c3b651.jpg
..\input\global-wheat-detection\train\c672ca0b0.jpg
..\input\global-wheat-detection\train\c682d8c66.jpg
..\input\global-wheat-detection\train\c6a299f4d.jpg
..\input\global-wheat-detection\train\c6b4933f7.jpg
..\input\global-wheat-detection\train\c6ca898c2.jpg
..\input\global-wheat-detection\train\c6d94be4c.jpg
..\input\global-wheat-detection\train\c6e6f5527.jpg
..\input\global-wheat-detection\train\c720c9774.jpg
..\input\global-wheat-detection\train\c7385941e.jpg
..\input\global-wheat-detection\train\c73f0a714.jpg
..\input\global-wheat-detection\train\c742fa9ae.jpg
..\input\global-wheat-detection\train\c74ca81bd.jpg
..\input\global-wheat-detection\train\c74db0380.jpg
..\input\global-wheat-detection\train\c7532e8e6.jpg
..\input\global-wheat-detection\train\c7637f1dd.jpg
..\input\glo

..\input\global-wheat-detection\train\e55ddec1a.jpg
..\input\global-wheat-detection\train\e596e6bb2.jpg
..\input\global-wheat-detection\train\e5ab88ff3.jpg
..\input\global-wheat-detection\train\e5b2763a1.jpg
..\input\global-wheat-detection\train\e5c0a16c5.jpg
..\input\global-wheat-detection\train\e5c45da08.jpg
..\input\global-wheat-detection\train\e5c9056ab.jpg
..\input\global-wheat-detection\train\e5e1f2c65.jpg
..\input\global-wheat-detection\train\e5f11f689.jpg
..\input\global-wheat-detection\train\e60108508.jpg
..\input\global-wheat-detection\train\e60d70a8f.jpg
..\input\global-wheat-detection\train\e6277ce7c.jpg
..\input\global-wheat-detection\train\e64e1c77b.jpg
..\input\global-wheat-detection\train\e67200e55.jpg
..\input\global-wheat-detection\train\e6989f6c7.jpg
..\input\global-wheat-detection\train\e69cbaece.jpg
..\input\global-wheat-detection\train\e6af2a27e.jpg
..\input\global-wheat-detection\train\e6af62317.jpg
..\input\global-wheat-detection\train\e6b54d480.jpg
..\input\glo

..\input\global-wheat-detection\train\fe342a5bf.jpg
..\input\global-wheat-detection\train\fe3d17ee2.jpg
..\input\global-wheat-detection\train\fe3f956a2.jpg
..\input\global-wheat-detection\train\fe5c1ea5a.jpg
..\input\global-wheat-detection\train\fe5c215e2.jpg
..\input\global-wheat-detection\train\fe641f02e.jpg
..\input\global-wheat-detection\train\fe6e2ecbc.jpg
..\input\global-wheat-detection\train\fe7ecb1f4.jpg
..\input\global-wheat-detection\train\fe8876602.jpg
..\input\global-wheat-detection\train\feac3a701.jpg
..\input\global-wheat-detection\train\feb09b731.jpg
..\input\global-wheat-detection\train\feb78a8de.jpg
..\input\global-wheat-detection\train\fec6edf64.jpg
..\input\global-wheat-detection\train\feda9265c.jpg
..\input\global-wheat-detection\train\fefd59985.jpg
..\input\global-wheat-detection\train\ff1716939.jpg
..\input\global-wheat-detection\train\ff46aa8d4.jpg
..\input\global-wheat-detection\train\ff5c9104a.jpg
..\input\global-wheat-detection\train\ff86547fb.jpg
..\input\glo

### Preprocessing

In [5]:
import pandas as pd
import numpy as np

#### Transforming the Labels for yolov5

yolov5 expects the labels in the form:

* one txt file per picture
* one line per object (Box)
* box coordinates have to be normalized (0-1) in the format: x_center, y_center, width, height
* class numbers (zero-indexed)

In [6]:
labels = pd.read_csv('../input/global-wheat-detection/train.csv')

In [7]:
labels

image_id  width  height                         bbox     source
0       b6ab77fd7   1024    1024   [834.0, 222.0, 56.0, 36.0]    usask_1
1       b6ab77fd7   1024    1024  [226.0, 548.0, 130.0, 58.0]    usask_1
2       b6ab77fd7   1024    1024  [377.0, 504.0, 74.0, 160.0]    usask_1
3       b6ab77fd7   1024    1024  [834.0, 95.0, 109.0, 107.0]    usask_1
4       b6ab77fd7   1024    1024  [26.0, 144.0, 124.0, 117.0]    usask_1
...           ...    ...     ...                          ...        ...
147788  5e0747034   1024    1024    [64.0, 619.0, 84.0, 95.0]  arvalis_2
147789  5e0747034   1024    1024  [292.0, 549.0, 107.0, 82.0]  arvalis_2
147790  5e0747034   1024    1024  [134.0, 228.0, 141.0, 71.0]  arvalis_2
147791  5e0747034   1024    1024   [430.0, 13.0, 184.0, 79.0]  arvalis_2
147792  5e0747034   1024    1024   [875.0, 740.0, 94.0, 61.0]  arvalis_2

[147793 rows x 5 columns]

The dataset contains the columns:
* `image_id`: the unique identifier for each of the pictures. There is a row for every box, so each picture can occur multiple times
* `width` & `height`: number of pixels of width and height of the pictures
* `bbox`: coordinates for the box. They are in pixels and follow the format \[x_min, y_min, width, height\]
* `source`: source of the picture. The pictures of the wheat were taken by different institutions across the world and the reason for the competition is that the models trained seem to work very well in one part of the world, but not generalize very well.

In [8]:
num_labels = np.shape(labels.image_id.unique())[0]
num_labels

3373

This means there should be about 3373 pictures. It should be a little more, because the explanation of the dataset states, that there are several pictures that do not contain any boxes.

In [9]:
num_pics = len(os.listdir("../input/global-wheat-detection/train"))
num_pics

3422

In [10]:
num_pics - num_labels

49

This means there are 49 pictures that do not contain any boxes. This is expected and not important when using yolov5, because it doesn't expect a .txt file for pictures that do not contain any boxes, so I won't have to make one

In [11]:
labels.bbox

0          [834.0, 222.0, 56.0, 36.0]
1         [226.0, 548.0, 130.0, 58.0]
2         [377.0, 504.0, 74.0, 160.0]
3         [834.0, 95.0, 109.0, 107.0]
4         [26.0, 144.0, 124.0, 117.0]
                     ...             
147788      [64.0, 619.0, 84.0, 95.0]
147789    [292.0, 549.0, 107.0, 82.0]
147790    [134.0, 228.0, 141.0, 71.0]
147791     [430.0, 13.0, 184.0, 79.0]
147792     [875.0, 740.0, 94.0, 61.0]
Name: bbox, Length: 147793, dtype: object

Now the individual pixel numbers need to be extracted. The individual lists are actually saved as lists, so I will use the pandas.Series.str.extract() method to extract the individual numbers to use for creating the txt files

In [12]:
pixels = labels.bbox.str.extract(r'(\d+\.*\d*)\,\s(\d+\.*\d*)\,\s(\d+\.*\d*)\,\s(\d+\.*\d*)')
pixels.columns = ['x_min', 'y_min', 'w_box', 'h_box']

In [13]:
labels = pd.concat([labels,pixels], axis = 1)

In [14]:
labels.head(5)

image_id  width  height                         bbox   source  x_min  \
0  b6ab77fd7   1024    1024   [834.0, 222.0, 56.0, 36.0]  usask_1  834.0   
1  b6ab77fd7   1024    1024  [226.0, 548.0, 130.0, 58.0]  usask_1  226.0   
2  b6ab77fd7   1024    1024  [377.0, 504.0, 74.0, 160.0]  usask_1  377.0   
3  b6ab77fd7   1024    1024  [834.0, 95.0, 109.0, 107.0]  usask_1  834.0   
4  b6ab77fd7   1024    1024  [26.0, 144.0, 124.0, 117.0]  usask_1   26.0   

   y_min  w_box  h_box  
0  222.0   56.0   36.0  
1  548.0  130.0   58.0  
2  504.0   74.0  160.0  
3   95.0  109.0  107.0  
4  144.0  124.0  117.0

In [15]:
labels.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 147793 entries, 0 to 147792
Data columns (total 9 columns):
 #   Column    Non-Null Count   Dtype 
---  ------    --------------   ----- 
 0   image_id  147793 non-null  object
 1   width     147793 non-null  int64 
 2   height    147793 non-null  int64 
 3   bbox      147793 non-null  object
 4   source    147793 non-null  object
 5   x_min     147793 non-null  object
 6   y_min     147793 non-null  object
 7   w_box     147793 non-null  object
 8   h_box     147793 non-null  object
dtypes: int64(2), object(7)
memory usage: 10.1+ MB


The `width`, `height`, `x_min`, `y_min`, `w_box` and `h_box` columns are still in a str format, but they can be easily transformed with the pandas.Series.astype('float64') method

In [16]:
labels.width = labels.width.astype('float64')
labels.height = labels.height.astype('float64')
labels.x_min = labels.x_min.astype('float64')
labels.y_min = labels.y_min.astype('float64')
labels.w_box = labels.w_box.astype('float64')
labels.h_box = labels.h_box.astype('float64')

In [17]:
labels.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 147793 entries, 0 to 147792
Data columns (total 9 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   image_id  147793 non-null  object 
 1   width     147793 non-null  float64
 2   height    147793 non-null  float64
 3   bbox      147793 non-null  object 
 4   source    147793 non-null  object 
 5   x_min     147793 non-null  float64
 6   y_min     147793 non-null  float64
 7   w_box     147793 non-null  float64
 8   h_box     147793 non-null  float64
dtypes: float64(6), object(3)
memory usage: 10.1+ MB


Now to make the normed coordinates that yolo needs:

In [18]:
labels['object_class'] = 0 # object class for use by yolo all objects are wheat so class 0
labels['x_center_norm'] = ((labels.x_min + 0.5 * labels.w_box) / labels.width).round(5)
labels['y_center_norm'] = ((labels.y_min + 0.5 * labels.h_box) / labels.height).round(5)
labels['width_norm'] = (labels.w_box/labels.width).round(5)
labels['height_norm'] = (labels.h_box/labels.height).round(5)

now the only thing left to do is to create the label .txt files.

In [19]:
id_nr = labels.image_id.unique()[0]
# np.savetxt(r'./labels/'+id_nr+r'.txt', labels[labels.image_id == id_nr].iloc[:, 9:].values, sep = ',')
for id_nr in labels.image_id.unique():
    labels[labels.image_id == id_nr].iloc[:, 9:].to_csv(r'./data/labels/train/'+id_nr+r'.txt',sep = ' ', index = False, header = False)

## Yolov5

The next step now will be to set up yolov5. To do this the data needs to be downloaded and the dependencies installed

In [20]:
!git clone https://github.com/jillerhaus/Working_Yolov5

Cloning into 'Working_Yolov5'...


In [21]:
if kaggle:
    !pip install -r yolov5/requirements.txt  # install dependencies
%cd Working_Yolov5
import torch
from IPython.display import Image, clear_output  # to display images
from utils.google_utils import gdrive_download  # to download models/datasets

clear_output()
print('Setup complete. Using torch %s %s' % (torch.__version__, torch.cuda.get_device_properties(0) if torch.cuda.is_available() else 'CPU'))

Setup complete. Using torch 1.5.1+cu101 _CudaDeviceProperties(name='GeForce RTX 2080 Ti', major=7, minor=5, total_memory=11264MB, multi_processor_count=68)


### Making the YAML File for Yolo

The solution I chose is pretty inellegant but I think it works alright

In [22]:
!echo # train and val datasets (image directory or *.txt file with image paths) > data/wheat.yaml
!echo train: ../data/images/train/ >> data/wheat.yaml
!echo val: ../data/images/train/ >> data/wheat.yaml
!echo # number of classes >> data/wheat.yaml
!echo nc: 1 >> data/wheat.yaml
!echo # class names >> data/wheat.yaml
!echo names: ['wheat'] >> data/wheat.yaml

In [23]:
if kaggle:
    !cat data/wheat.yaml

### Training Yolo

In this step a Yolov5s model is trained on the dataset

In [24]:
if windows:
    !copy ..\..\input\yolos\* .

..\..\input\yolos\best.pt
..\..\input\yolos\best_tutorial.pt
..\..\input\yolos\last.pt
..\..\input\yolos\last_tutorial.pt
..\..\input\yolos\yolov3-spp.pt
..\..\input\yolos\yolov5l.pt
..\..\input\yolos\yolov5m.pt
..\..\input\yolos\yolov5s.pt
..\..\input\yolos\yolov5x.pt
        9 file(s) copied.


In [25]:
import apex

In [27]:
!python train.py --img 1024 --batch 4 --epochs 10 --data ./data/wheat.yaml --cfg ./models/yolov5x.yaml --weights last_tutorial.pt --name tutorial

{'lr0': 0.01, 'momentum': 0.937, 'weight_decay': 0.0005, 'giou': 0.05, 'cls': 0.58, 'cls_pw': 1.0, 'obj': 1.0, 'obj_pw': 1.0, 'iou_t': 0.2, 'anchor_t': 4.0, 'fl_gamma': 0.0, 'hsv_h': 0.014, 'hsv_s': 0.68, 'hsv_v': 0.36, 'degrees': 0.0, 'translate': 0.0, 'scale': 0.5, 'shear': 0.0}
{'lr0': 0.01, 'momentum': 0.937, 'weight_decay': 0.0005, 'giou': 0.05, 'cls': 0.58, 'cls_pw': 1.0, 'obj': 1.0, 'obj_pw': 1.0, 'iou_t': 0.2, 'anchor_t': 4.0, 'fl_gamma': 0.0, 'hsv_h': 0.014, 'hsv_s': 0.68, 'hsv_v': 0.36, 'degrees': 0.0, 'translate': 0.0, 'scale': 0.5, 'shear': 0.0}
{'lr0': 0.01, 'momentum': 0.937, 'weight_decay': 0.0005, 'giou': 0.05, 'cls': 0.58, 'cls_pw': 1.0, 'obj': 1.0, 'obj_pw': 1.0, 'iou_t': 0.2, 'anchor_t': 4.0, 'fl_gamma': 0.0, 'hsv_h': 0.014, 'hsv_s': 0.68, 'hsv_v': 0.36, 'degrees': 0.0, 'translate': 0.0, 'scale': 0.5, 'shear': 0.0}
{'lr0': 0.01, 'momentum': 0.937, 'weight_decay': 0.0005, 'giou': 0.05, 'cls': 0.58, 'cls_pw': 1.0, 'obj': 1.0, 'obj_pw': 1.0, 'iou_t': 0.2, 'anchor_t': 4.

2020-07-27 14:55:15.866233: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudart64_101.dll

  0%|          | 0/3422 [00:00<?, ?it/s]
Caching labels ..\data\labels\train.npy (2105 found, 0 missing, 27 empty, 0 duplicate, for 3422 images):  62%|######2   | 2132/3422 [00:00<00:00, 21108.70it/s]
Caching labels ..\data\labels\train.npy (3373 found, 0 missing, 49 empty, 0 duplicate, for 3422 images): 100%|##########| 3422/3422 [00:00<00:00, 21254.63it/s]

  0%|          | 0/3422 [00:00<?, ?it/s]
Caching labels ..\data\labels\train.npy (2103 found, 0 missing, 27 empty, 0 duplicate, for 3422 images):  62%|######2   | 2130/3422 [00:00<00:00, 21299.72it/s]
Caching labels ..\data\labels\train.npy (3373 found, 0 missing, 49 empty, 0 duplicate, for 3422 images): 100%|##########| 3422/3422 [00:00<00:00, 21387.36it/s]

  0%|          | 0/856 [00:00<?, ?it/s]
     48/56     7.89G   0.02809    0.1033         0    0.1314       209      1024:   0%|   

{'lr0': 0.01, 'momentum': 0.937, 'weight_decay': 0.0005, 'giou': 0.05, 'cls': 0.58, 'cls_pw': 1.0, 'obj': 1.0, 'obj_pw': 1.0, 'iou_t': 0.2, 'anchor_t': 4.0, 'fl_gamma': 0.0, 'hsv_h': 0.014, 'hsv_s': 0.68, 'hsv_v': 0.36, 'degrees': 0.0, 'translate': 0.0, 'scale': 0.5, 'shear': 0.0}
{'lr0': 0.01, 'momentum': 0.937, 'weight_decay': 0.0005, 'giou': 0.05, 'cls': 0.58, 'cls_pw': 1.0, 'obj': 1.0, 'obj_pw': 1.0, 'iou_t': 0.2, 'anchor_t': 4.0, 'fl_gamma': 0.0, 'hsv_h': 0.014, 'hsv_s': 0.68, 'hsv_v': 0.36, 'degrees': 0.0, 'translate': 0.0, 'scale': 0.5, 'shear': 0.0}
{'lr0': 0.01, 'momentum': 0.937, 'weight_decay': 0.0005, 'giou': 0.05, 'cls': 0.58, 'cls_pw': 1.0, 'obj': 1.0, 'obj_pw': 1.0, 'iou_t': 0.2, 'anchor_t': 4.0, 'fl_gamma': 0.0, 'hsv_h': 0.014, 'hsv_s': 0.68, 'hsv_v': 0.36, 'degrees': 0.0, 'translate': 0.0, 'scale': 0.5, 'shear': 0.0}
{'lr0': 0.01, 'momentum': 0.937, 'weight_decay': 0.0005, 'giou': 0.05, 'cls': 0.58, 'cls_pw': 1.0, 'obj': 1.0, 'obj_pw': 1.0, 'iou_t': 0.2, 'anchor_t': 4.

     48/56     8.15G   0.03223    0.1172         0    0.1494       175      1024:   4%|3         | 34/856 [00:24<08:00,  1.71it/s]
     48/56     8.15G   0.03223    0.1172         0    0.1494       175      1024:   4%|4         | 35/856 [00:24<07:47,  1.76it/s]
     48/56     8.15G   0.03237    0.1173         0    0.1497       177      1024:   4%|4         | 35/856 [00:24<07:47,  1.76it/s]
     48/56     8.15G   0.03237    0.1173         0    0.1497       177      1024:   4%|4         | 36/856 [00:24<07:40,  1.78it/s]
     48/56     8.15G   0.03238    0.1167         0    0.1491       129      1024:   4%|4         | 36/856 [00:25<07:40,  1.78it/s]
     48/56     8.15G   0.03238    0.1167         0    0.1491       129      1024:   4%|4         | 37/856 [00:25<07:36,  1.79it/s]
     48/56     8.15G    0.0325    0.1173         0    0.1498       208      1024:   4%|4         | 37/856 [00:26<07:36,  1.79it/s]
     48/56     8.15G    0.0325    0.1173         0    0.1498       208      1024:  